sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -118,128 +144,184 @@ private void drainQueue() {
-                    currentWork.startedProcessing();
-                    final long calculatedRequest = currentWork.getNumberOfEvents() - remaining.get();
-                    remaining.addAndGet(calculatedRequest);
-                    subscription.request(calculatedRequest);
+                    isEmitted = work.emitNext(message);
                 }
 
-                // Send messages to currentWork from buffer
-                while (bufferMessages.size() > 0 && !currentWork.isTerminal()) {
-                    currentWork.next(bufferMessages.poll());
-                    remaining.decrementAndGet();
+                // We could not emit the last message that we polled because there were no work items.
+                // Push this back to the head of the work queue.
+                if (!isEmitted) {
+                    bufferMessages.addFirst(message);
+                    break;
                 }
 
-                // if  we have delivered all the messages to currentWork, we will complete it.
-                if (currentWork.isTerminal()) {
-                    if (currentWork.getError() == null) {
-                        currentWork.complete();
-                    }
-                    // Now remove from queue since it is complete
-                    workQueue.remove(currentWork);
-                    if (currentTimeoutOperation != null && !currentTimeoutOperation.isDisposed()) {
-                        currentTimeoutOperation.dispose();
-                    }
-                    logger.verbose("The work [{}] is complete.", currentWork.getId());
-                }
+                numberEmitted++;
+                isEmpty = bufferMessages.isEmpty();
             }
-        }
-    }
 
-    /**
-     * @param work on which timeout thread need to start.
-     *
-     * @return {@link Disposable} for the timeout operation.
-     */
-    private Disposable getTimeoutOperation(SynchronousReceiveWork work) {
-        Duration timeout = work.getTimeout();
-        return Mono.delay(timeout).thenReturn(work)
-            .subscribe(l -> {
-                synchronized (currentWorkLock) {
-                    if (currentWork == work) {
-                        work.timeout();
-                    }
-                }
-            });
+            final long requestedMessages = REQUESTED.get(this);
+            if (requestedMessages != Long.MAX_VALUE) {
+                numberRequested = REQUESTED.addAndGet(this, -numberEmitted);
+            }
+        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     protected void hookOnError(Throwable throwable) {
-        logger.error("[{}] Errors occurred upstream", currentWork.getId(), throwable);
+        logger.error("workId[{}] Errors occurred upstream", currentWork.getId(), throwable);
         synchronized (currentWorkLock) {
-            currentWork.error(throwable);
+            currentWork.complete(null, throwable);
```

Comment:
When an error (terminal event), we are completing the "current-work." I wonder what would happen to the pending works already in the `this.workQueue`. Are we completing each of them? such that any attempt to iterate on Iterator those works backs not hangs but terminate?

Reply:
Ah yes. I caught that while writing additional test cases. I have resolved this.